experimental/agent/gen_agent.py

```diff
@@ -0,0 +1,164 @@
+        threshold=generative_models.HarmBlockThreshold.BLOCK_ONLY_HIGH,
+    ),
+    generative_models.SafetySetting(
+        category=generative_models.HarmCategory.HARM_CATEGORY_HATE_SPEECH,
+        threshold=generative_models.HarmBlockThreshold.BLOCK_ONLY_HIGH,
+    ),
+    generative_models.SafetySetting(
+        category=generative_models.HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
+        threshold=generative_models.HarmBlockThreshold.BLOCK_ONLY_HIGH,
+    ),
+]
+
+
+def _start_docker_container() -> str:
+  result = subprocess.run(
+      ['docker', 'run', '-d', '-t', '--entrypoint=/bin/bash', IMAGE_NAME],
+      stdout=subprocess.PIPE,
+      stderr=subprocess.PIPE,
+      text=True,
+      check=True)
+  container_id = result.stdout.strip()
+  return container_id
+
+
+def _construct_prompt(content: str = '') -> str:
+  with open('prompts/agent/gen-initial.txt', 'r') as prompt_file:
+    content = prompt_file.read()
+    content = content.replace('{FUZZ_TARGET_PATH}', FUZZ_TARGET_PATH)
+    content = content.replace('{PROJECT}', PROJECT)
+    content = content.replace('{FUNCTION_UNDER_TEST}', FUNCTION_UNDER_TEST)
+    return content
+
+
+def _parse_tag(response: str, tag: str) -> str:
+  match = re.search(rf'<{tag}>(.*?)</{tag}>', response, re.DOTALL)
+  return match.group(1).strip() if match else ''
+
+
+def execute_bash_command(container_id: str, command: str) -> str:
+  result = subprocess.run(
+      ['docker', 'exec', container_id, '/bin/bash', '-c', command],
+      stdout=subprocess.PIPE,
+      stderr=subprocess.PIPE,
+      check=False,
+      text=True)
+  print('ERR', result.stderr)
+  print('OUT', result.stdout)
+  print('RET', result.returncode)
+  print('FIN', result.stderr if result.returncode else result.stdout)
+  # return result.stdout if result.returncode == 0 else result.stderr
+  return result.stderr if result.stderr else result.stdout
+
+
+def main() -> None:
+  chat = model.start_chat()
+
+  # Start the Docker container
+  container_id = _start_docker_container()
+  print(f"Started Docker container: {container_id}")
+  # fuzz_target_basename, ext = os.path.splitext(FUZZ_TARGET_PATH)
+  # example = f'{fuzz_target_basename}_example{ext}'
+  # execute_bash_command(container_id, f'mv "{FUZZ_TARGET_PATH}" "{example}"')
+
+  # replace_target = f'echo "{CODE_TO_BE_FIXED}" > "{FUZZ_TARGET_PATH}"'
+  # execute_bash_command(container_id, replace_target)
+  execute_bash_command(
+      container_id,
+      'compile > /src/compile_output && rm -rf /out/* > /dev/null')
+  output = ''
+  response = ''
+  cur_round = 0
+  prompt = _construct_prompt(output)
+
+  try:
+    while cur_round < 100:
+      cur_round += 1
+      print(f"ROUND {cur_round} Prompt:\n{prompt}")
+      response = chat.send_message(
+          prompt or ' ', stream=False,
+          safety_settings=SAFETY_CONFIG).text  # type: ignore
+
+      print(f"ROUND {cur_round} LLM response:\n{response}")
+      if _parse_tag(response, 'conclusion'):
```

Comment:
Could we start refactoring all of this into a more generalized framework that we can reuse and extend? e.g. something along the lines of:

```python
class AgentPromptSession:
  initial_prompt: str

  def add_response_handler(tag_name: str, action: Callable):
    ...
   

class FixerPromptSession(AgentPromptSession):
  def __init__(self):
    self.initial_prompt = ...
    self.add_response_handler('conclusion', self._handle_conclusion)
    self.add_response_handler('bash', self._handle_bash)
```

(just first pass at a suggestion above -- it's might not be the best way to structure this).

Reply:
Yep, thanks!
An `lldb-agent` is also planned and will be added soon.
I will follow this when designing the framework for them all.